pkg/ccl/backupccl/file_sst_sink_test.go

```diff
@@ -135,3 +141,819 @@ func TestFileSSTSinkExtendOneFile(t *testing.T) {
+				if input.expectErr != "" {
+					// Do not compare resulting SSTs if we expect errors.
+					require.ErrorContains(t, err, input.expectErr)
+					compareSST = false
+				} else {
+					require.NoError(t, err)
+				}
+			}
+
+			sst.Close()
+
+			if !compareSST {
+				return
+			}
+
+			var expected []kvAndTS
+			for _, input := range tt.inputs {
+				expected = append(expected, input.input...)
+			}
+
+			iterOpts := storage.IterOptions{
+				KeyTypes:   storage.IterKeyTypePointsAndRanges,
+				LowerBound: keys.LocalMax,
+				UpperBound: keys.MaxKey,
+			}
+
+			iter, err := storage.NewMemSSTIterator(buf.Bytes(), false, iterOpts)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			var actual []kvAndTS
+			for iter.SeekGE(storage.MVCCKey{}); ; iter.Next() {
+				if ok, err := iter.Valid(); err != nil {
+					t.Fatal(err)
+				} else if !ok {
+					break
+				}
+
+				key := iter.UnsafeKey()
+				value, err := iter.UnsafeValue()
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				kv := kvAndTS{key: string(key.Key), timestamp: key.Timestamp.WallTime}
+				kv.value = append(kv.value, value...)
+
+				actual = append(actual, kv)
+			}
+
+			require.Equal(t, expected, actual)
+		})
+	}
+}
+
+func TestFileSSTSinkCopyRangeKeys(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	settings := cluster.MakeTestingClusterSettings()
+
+	type testInput struct {
+		input     []rangeKeyAndTS
+		expectErr string
+	}
+
+	type testCase struct {
+		name   string
+		inputs []testInput
+	}
+
+	for _, tt := range []testCase{
+		{
+			name: "single-sst",
+			inputs: []testInput{
+				{
+					input: []rangeKeyAndTS{
+						{key: "a", endKey: "b", timestamp: 10},
+						{key: "b", endKey: "c", timestamp: 9},
+					},
+				},
+			},
+		},
+		{
+			name: "multiple-ssts",
+			inputs: []testInput{
+				{
+					input: []rangeKeyAndTS{
+						{key: "a", endKey: "b", timestamp: 10},
+						{key: "b", endKey: "c", timestamp: 9},
+					},
+				},
+				{
+					input: []rangeKeyAndTS{
+						{key: "c", endKey: "d", timestamp: 10},
+						{key: "c", endKey: "d", timestamp: 9},
+					},
+				},
+				{
+					input: []rangeKeyAndTS{
+						{key: "c", endKey: "d", timestamp: 8},
+					},
+				},
+			},
+		},
+		{
+			name: "out-of-order-range",
+			inputs: []testInput{
+				{
+					input: []rangeKeyAndTS{
+						{key: "a", endKey: "b", timestamp: 10},
+						{key: "b", endKey: "d", timestamp: 9},
+					},
+				},
+				{
+					input: []rangeKeyAndTS{
+						{key: "a", endKey: "c", timestamp: 8},
+						{key: "c", endKey: "e", timestamp: 7},
+					},
+					expectErr: "spans must be added in order",
+				},
+			},
+		},
+		{
+			name: "out-of-order-timestamp",
+			inputs: []testInput{
+				{
+					input: []rangeKeyAndTS{
+						{key: "a", endKey: "b", timestamp: 10},
+						{key: "b", endKey: "d", timestamp: 9},
+					},
+				},
+				{
+					input: []rangeKeyAndTS{
+						{key: "b", endKey: "d", timestamp: 9},
```

Comment:
it looks like the rangkeys themselves are out of order? i'm kinda surprised this passes?

Reply:
yeah I was surprised as well. I looked at the docstring of the pebble fragmenter and it doesn't mention anything about needing the timestamp in order: https://github.com/cockroachdb/pebble/blob/master/internal/keyspan/fragmenter.go#L106-L109

Reply:
Aha, but what about this [line](https://github.com/cockroachdb/pebble/blob/master/internal/keyspan/fragmenter.go#L126):

> // The fragments need to be output sorted by start key, and for equal start
// keys, sorted by descending sequence number.

I'm pretty sure that means that on the pebble level, range keys with the same start key are sorted by some notion of time (sequence number), for newest to oldest key i think.

Reply:
yeah I just meant that it seems that even if the ranges are supplied to `sst.PutRawMVCCRangeKey()` out of timestamp order, the writer still does the correct thing. It seems that the requirement for ordering only applies to the start key of the spans of the inputs, but not the timestamps.

Reply:
I changed the test a little so that there should be unique entries for the two different times of range `[b, d)`. The test seems to still pass and I manually verified that all of the versions were copied correctly.

Reply:
I verified that the SST contains entries for the rangekey for all timestamps if the original SSTs had entries for the same range with multiple timestamps.